Add docs for controllable parameters API (#73734)
metabase/metabase: docs/embedding/sdk/questions.md

{% include_file "{{ dirname }}/api/snippets/InteractiveQuestionProps.md" snippet="properties" %}

> ## Pass SQL parameters to SQL questions

Reviewer: ```suggestion
## Pass parameters to SQL questions
```

You can pass parameter values to questions defined with SQL in the format of `{parameter_name: parameter_value}`. Learn more about [SQL parameters](../../questions/native-editor/sql-parameters.md).

There are two ways to pass them - pick one depending on whether you only want to seed initial values or stay in sync with applied changes over time.

> These props can't be used with questions built using the query builder.

Reviewer: ```suggestion
You can pass [parameter](../../questions/native-editor/sql-parameters.md) values to SQL questions like so:

```
{parameter_name: parameter_value}
```

There are two ways to pass parameters:

- Uncontrolled, where the component owns the state: use `initialSqlParameters` to set initial values.
- Controlled, where your app owns the parameter state. Use `sqlParameters` and `onSqlParametersChange` to sync the parameter values with your app when people viewing the embed change the filters.

These props can only be used with SQL questions; they don't work for query-builder questions.
```

> ### `initialSqlParameters` (uncontrolled)

Reviewer: The `initialSqlParameters` is an implementation detail.

```suggestion
### Uncontrolled component with `initialSqlParameters`
```

> Pre-populate parameters once on load. After the question mounts, user widget edits are not reflected back to your app.

Reviewer: ```suggestion
Sets the parameter value once on load. After the question mounts, if the person viewing the embed changes the value, your app won't know about it.
```

```typescript
{% include_file "{{ dirname }}/snippets/questions/initial-sql-parameters.tsx" snippet="example" %}
```

### `sqlParameters` + `onSqlParametersChange` (controlled)

Push values from your app, and observe every applied change via `onSqlParametersChange`. This works like a controlled `<input value onChange>` - your app holds the source of truth, the question re-renders when the prop changes, and you receive a callback whenever applied values change.

> 

Reviewer: ```suggestion
### Controlled component with `sqlParameters` and `onSqlParametersChange`

With this setup, your app maintains the parameter value state. When someone viewing the embed updates the value, a callback fires to update your app with the new value, and the question re-renders with this new value.  This works like a controlled `<input value onChange>`.
```

```typescript
{% include_file "{{ dirname }}/snippets/questions/controlled-sql-parameters.tsx" snippet="example-controlled" %}
```

The callback receives a payload with these fields:

- `source` - why the callback fired:
  - `"initial-state"` - the post-load snapshot, fired once per question load.
  - `"manual-change"` - a user edited a filter widget.
  - `"auto-change"` - values you pushed via `sqlParameters` were stored differently than you sent (for example, a scalar wrapped into an array for a multi-select filter, an unknown slug ignored, an explicit `null` applied as a clear). Use the payload to re-sync your local state.
- `parameters` - the currently applied values, slug-keyed.

> - `defaultParameters` - the question's default values, slug-keyed.

Reviewer: ```suggestion
- `source`: why the callback fired:
  - `"initial-state"`: the post-load snapshot, fired once per question load.
  - `"manual-change"`: someone edited a filter widget.
  - `"auto-change"`: values you pushed via `sqlParameters` were stored differently than you sent. For example, a scalar wrapped into an array for a multi-select filter, an unknown slug ignored, an explicit `null` applied as a clear. Use the payload to re-sync the local state in your app.
- `parameters` : the currently applied values, slug-keyed.
- `defaultParameters`: the question's default values, slug-keyed.
```

> If your push is applied unchanged, no callback fires.

Reviewer: > If your push is applied unchanged, no callback fires.

Meaning: if the parameter value you send to the app is the same as the existing value, nothing happens?

#### Clearing a parameter

> Set its value to `null`. This strictly clears the parameter and ignores its default value. Missing slugs fall back to `parameter.default ?? null`.

Reviewer: ```suggestion
To clear a parameter completely, set its value to `null`. Clearing the parameter with a `null` also ignores the parameter's default value (if any). Missing slugs fall back to `parameter.default ?? null`.
```

```typescript
{% include_file "{{ dirname }}/snippets/questions/controlled-sql-parameters.tsx" snippet="example-clear" %}
```

> To switch back to uncontrolled mode (and leave the question's last applied values in place), set the `sqlParameters` prop to `undefined`.

Reviewer: ```suggestion
To switch to uncontrolled mode without clearing the value, set the `sqlParameters` prop to `undefined`. The value will remain whatever it was last set to. Your app won't get any further updates when people update the value for this parameter.
```

Does this edit have it right?

Author: No, actually i'd remove this part completely as in real scenarios most likely nobody will do it

> > Don't combine `initialSqlParameters` and `sqlParameters` - pick one. For controlled behavior, only use `sqlParameters`.

Reviewer: This shouldn't be a block quote. Can we move this warning to the beginning of the section when we introduce the two approaches?
